agent-nodejs/src/opamp/client-http.ts

```diff
@@ -0,0 +1,208 @@
+
+      // flags is bitmap, use bitwise OR to check if the flag is set
+      if (
+        Number(heartbeatRes.flags) |
+        Number(ServerToAgentFlags.ServerToAgentFlags_ReportFullState)
+      ) {
+        this.logger.info("Opamp server requested full state report");
+        heartbeatRes = await this.sendFullState();
+      }
+    }, this.config.pollingIntervalMs || 30000);
+    timer.unref(); // do not keep the process alive just for this timer
+  }
+
+  async shutdown() {
+    this.logger.info("Sending AgentDisconnect message to OpAMP server");
+    return await this.sendAgentToServerMessage({
+      agentDisconnect: {},
+    });
+  }
+
+  // the first opamp message is special, as we need to get the remote resource attributes.
+  // this function will attempt to send the first message, and will retry after some interval if it fails.
+  // if no remote resource attributes are received after some grace period, we will continue without them.
+  private async sendFirstMessageWithRetry() {
+    let remainingRetries = 5;
+    const retryIntervalMs = 2000;
+
+    for (let i = 0; i < remainingRetries; i++) {
+      try {
+        const firstServerToAgent = await this.sendFullState();
+
+        const sdkConfig =
+          firstServerToAgent.remoteConfig?.config?.configMap["SDK"];
+        if (!sdkConfig || !sdkConfig.body) {
+          throw new Error(
+            "No SDK config received on first OpAMP message to the server"
+          );
+        }
+
+        const resourceAttributes = JSON.parse(sdkConfig.body.toString()) as {
+          remoteResourceAttributes: ResourceAttributeFromServer[];
+        };
+
+        if (this.resourcePromiseResolver) {
+          this.logger.info(
+            "Got remote resource attributes, resolving detector promise",
+            resourceAttributes.remoteResourceAttributes
+          );
+          this.resourcePromiseResolver(
+            keyValuePairsToOtelAttributes([
+              ...resourceAttributes.remoteResourceAttributes,
+              {
+                key: SEMRESATTRS_SERVICE_INSTANCE_ID,
+                value: this.OpAMPInstanceUidString,
+              },
+            ])
+          );
+          return;
+        }
+      } catch (error) {
+        this.logger.warn(
+          `Error sending first message to OpAMP server, retrying in ${retryIntervalMs}ms`,
+          error
+        );
+        await new Promise((resolve) => {
+          const timer = setTimeout(resolve, retryIntervalMs);
+          timer.unref(); // do not keep the process alive just for this timer
+        });
+      }
+    }
+
+    // if we got here, it means we run out of retries and did not return from the loop
+    this.logger.error(
+      `Failed to get remote resource attributes from OpAMP server after retries, continuing without them`
+    );
+    this.resourcePromiseResolver?.({
+      [SEMRESATTRS_SERVICE_NAME]: this.config.instrumentationDeviceId,
+    });
+  }
+
+  private async sendHeartBeatToServer() {
```

Comment:
Do we need the heartbeat to keep the connection alive or is it part of OpAMP spec? Maybe we can drop it if we are not doing anything on the heartbeat

Reply:
OpAMP server uses it to figure out if the process (or the connection) is down